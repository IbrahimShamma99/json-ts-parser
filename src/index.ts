export { Scanner } from './scanner'
export { Parser } from './parser'
