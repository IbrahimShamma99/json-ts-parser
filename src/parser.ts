import { ArrayExpression } from './expressions/array.expression'
import { ObjectExpression } from './expressions/object.expression'
import { Token, TokenType } from './types'

export class Parser {
  public tree: (ObjectExpression | ArrayExpression)[] = []

  private current: number = 0

  private readonly tokens!: Token[]

  constructor(tokens: Token[]) {
    this.tokens = tokens
  }

  parse() {
    while (!this.isAtEnd) {
      // this.tree.push(this.parseValue())
    }
  }

  private parseValue() {
    if (this.match(TokenType.LEFT_BRACKET)) {
      return this.parseObject()
    } else if (this.match(TokenType.SQUARE_LEFT_BRACKET)) {
      return this.parseArray()
    }
  }

  private parseArray() {}

  private parseObject() {
    // const object = new ObjectExpression()
    // while (!this.match(TokenType.RIGHT_BRACKET)) {
    //   object.add(this.parseObjectMember())
    // }
    // return object
  }

  private match(...tokens: TokenType[]) {
    if (this.check(...tokens)) {
      this.advance()
      return true
    }
    return false
  }

  private check(...tokens: TokenType[]): boolean {
    return tokens.includes(this.currentToken.type)
  }

  private get currentToken() {
    return this.tokens[this.current]
  }

  advance() {
    if (!this.isAtEnd) {
      this.current++
    }
    return this.previous
  }

  get isAtEnd() {
    return this.peek.type === TokenType.EOF
  }

  get peek() {
    return this.tokens[this.current]
  }

  get previous() {
    return this.tokens[this.current - 1]
  }
}
