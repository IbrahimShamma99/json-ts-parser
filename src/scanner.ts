import { Token, TokenType } from './types'

export class Scanner {
  public tokens: Token[] = []
  private current: number = 0
  private start: number = 0
  private line: number = 1

  private scanningCode!: string

  constructor(private readonly sourceCode: string) {
    this.preprocessSourceCode()
  }

  scan() {
    while (!this.isAtEnd) {
      this.start = this.current
      this.scanTokens()
      this.advance()
    }
    this.addEOF()
  }

  scanTokens() {
    const char = this.peek
    switch (char) {
      case ' ':
      case '\r':
      case '\t':
        break
      case '\n':
        this.line++
        break
      case 't':
        while (this.isCurrentChar) {
          this.advance()
        }
        if (this.lexeme === 'true') {
          this.addToken(TokenType.TRUE)
        } else {
          throw new Error('Invalid token')
        }
        break
      case 'n':
        while (this.isCurrentChar) {
          this.advance()
        }
        if (this.lexeme === 'null') {
          this.addToken(TokenType.NULL)
        } else {
          throw new Error('Invalid token')
        }
        break
      case 'f':
        while (this.isCurrentChar) {
          this.advance()
        }
        if (this.lexeme === 'false') {
          this.addToken(TokenType.FALSE)
        } else {
          throw new Error('Invalid token')
        }
        break
      case '[':
        this.addToken(TokenType.SQUARE_LEFT_BRACKET, '[')
        break
      case ']':
        this.addToken(TokenType.SQUARE_RIGHT_BRACKET, ']')
        break
      case '{':
        this.addToken(TokenType.LEFT_BRACKET, '{')
        break
      case '}':
        this.addToken(TokenType.RIGHT_BRACKET, '}')
        break
      case ':':
        this.addToken(TokenType.COLON, ':')
        break
      case ',':
        this.addToken(TokenType.COMMA, ',')
        break
      case '.':
        this.addToken(TokenType.DOT, '.')
        break
      case '"':
        this.string()
        break
      default:
        if (this.isCurrentNumber) {
          this.number()
          break
        }
        throw new Error('Invalid Token')
    }
  }

  get peek(): string {
    return this.scanningCode[this.current]
  }

  get peekNext(): string {
    return this.scanningCode[this.current + 1]
  }

  get isAtEnd(): boolean {
    return this.current >= this.scanningCode.length
  }

  get currentToken(): Token {
    return this.tokens[this.current - 1]
  }

  get lexeme() {
    return this.scanningCode.substring(this.start, this.current)
  }

  get isCurrentNumber() {
    return this.peek >= '0' && this.peek <= '9'
  }

  get isCurrentChar() {
    return (
      (this.peek >= 'a' && this.peek <= 'z') ||
      (this.peek >= 'A' && this.peek <= 'Z')
    )
  }

  match(lexeme: string) {
    if (this.peek === lexeme) {
      this.advance()
      return true
    }
    return false
  }

  advance() {
    this.current++
  }

  addToken(type: TokenType, initLexeme?: string) {
    this.tokens.push({
      lexeme: initLexeme ?? this.lexeme,
      line: this.line,
      end: this.current,
      start: this.start,
      type: type,
    })
  }

  addEOF() {
    this.tokens.push(
      new Token({
        lexeme: '\0',
        end: this.scanningCode.length,
        start: this.scanningCode.length,
        line: this.line,
        type: TokenType.EOF,
      })
    )
  }

  string() {
    this.advance()
    while (this.peek !== '"') {
      this.advance()
    }
    if (this.peekNext === ':') {
      this.addToken(
        TokenType.IDENTIFIER,
        this.scanningCode.substring(this.start + 1, this.current)
      )
    } else {
      this.addToken(
        TokenType.STRING,
        this.scanningCode.substring(this.start + 1, this.current)
      )
    }
  }

  number() {
    while (this.isCurrentNumber) {
      this.advance()
    }
    if (this.peek === TokenType.DOT) {
      while (this.isCurrentNumber) {
        this.advance()
      }
    }
    this.addToken(TokenType.NUMBER)
    if (this.peek === ',') {
      this.start = this.current
      this.advance()
      this.addToken(TokenType.COMMA)
    }
  }

  preprocessSourceCode() {
    this.scanningCode = this.sourceCode
      .replace(/{/g, ' { ')
      .replace(/}/g, ' } ')
  }
}
