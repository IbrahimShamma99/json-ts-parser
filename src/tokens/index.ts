import { TokenType } from './tokentype'
export * from './tokentype'

export class Token<T extends TokenType> {
  public readonly type: T
  public readonly lexeme: string
  public readonly line: number
  public readonly start: number
  public readonly end: number

  constructor({
    type,
    lexeme,
    line,
    start,
    end,
  }: {
    type: T
    lexeme: string
    line: number
    start: number
    end: number
  }) {
    this.type = type
    this.lexeme = lexeme
    this.line = line
    this.start = start
    this.end = end
  }
}
