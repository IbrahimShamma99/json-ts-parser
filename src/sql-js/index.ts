import { JsVisitor } from "../interpreter/js.visitor";
import { Scanner } from "../scanner";
import { Parser } from "../parser";

export const generate = (data: Record<any, any>, sql: string) => {
  const scanner = new Scanner(sql);
  scanner.scan();

  const parser = new Parser(scanner.tokens);

  const tree = parser.parse();

  const jsVisitor = new JsVisitor().execute(tree);

  console.log(jsVisitor);

  const jsCode = `function run() { return ${JSON.stringify(
    data
  )}${jsVisitor} } run()`;

  console.log(jsCode);

  return eval(jsCode);
};
