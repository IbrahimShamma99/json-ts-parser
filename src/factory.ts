import { Expression } from "./classes/expressions/expression";
import {
  UnaryExpression,
  UnaryToken,
} from "./classes/expressions/unary.expression";
import { Identifier } from "./classes/identifiers/identifier";
import { NumericLiteral } from "./classes/literals/numeric.literal";
import { SelectStatement } from "./classes/statements/select.statement";
import { TokenType, Token } from "./tokens";

export class Factory {
  public createSelectStatement(
    columns: (Expression | Identifier)[],
    from?: Expression,
    where?: Expression,
    order?: Expression,
    limit?: Expression
  ): SelectStatement {
    const statement = new SelectStatement();
    statement.columns = columns;
    statement.from = from;
    statement.where = where;
    statement.order = order;
    statement.limit = limit;
    return statement;
  }

  public createIdentifier(name: string, alias?: string): Identifier {
    return new Identifier(name, alias);
  }

  public createNumericLiteral(value: string): NumericLiteral {
    const numericLiteral = new NumericLiteral(value);
    return numericLiteral;
  }

  public createUnaryExpression(
    operator: Token<UnaryToken>,
    right: Expression
  ): UnaryExpression {
    return new UnaryExpression(operator, right);
  }
}
export function getKeyByValue(object: Record<string, any>, value: TokenType) {
  return Object.keys(object).find((key) => object[key] === value);
}
export const keywords: Record<string, TokenType> = {
  as: TokenType.AS,
  by: TokenType.BY,
  select: TokenType.SELECT,
  view: TokenType.VIEW,
  create: TokenType.CREATE,
  table: TokenType.TABLE,
  integer: TokenType.INTEGER,
  int: TokenType.INTEGER,
  text: TokenType.TEXT,
  real: TokenType.REAL,
  bool: TokenType.BOOL,
  boolean: TokenType.BOOL,
  date: TokenType.DATE,
  blob: TokenType.BLOB,
  unique: TokenType.UNIQUE,
  primary: TokenType.PRIMARY,
  key: TokenType.KEY,
  temp: TokenType.TEMP,
  temporary: TokenType.TEMP,
  check: TokenType.CHECK,
  default: TokenType.DEFAULT,
  from: TokenType.FROM,
  desc: TokenType.DESC,
  asc: TokenType.ASC,
  order: TokenType.ORDER,
  group: TokenType.GROUP,
  having: TokenType.HAVING,
  where: TokenType.WHERE,
  and: TokenType.AND,
  or: TokenType.OR,
  is: TokenType.IS,
  not: TokenType.NOT,
  "is not": TokenType.IS_NOT,
  "not between": TokenType.NOT_BETWEEN,
  "not ilike": TokenType.NOT_ILIKE,
  "not like": TokenType.NOT_LIKE,
  all: TokenType.ALL,
  distinct: TokenType.DISTINCT,
  limit: TokenType.LIMIT,
  offset: TokenType.OFFSET,
  like: TokenType.LIKE,
  ilike: TokenType.ILIKE,
  null: TokenType.NULL,
  in: TokenType.IN,
  similar: TokenType.SIMILAR,
  between: TokenType.BETWEEN,
  true: TokenType.FALSE,
  false: TokenType.TRUE,
  exists: TokenType.EXISTS,
  join: TokenType.JOIN,
  cross: TokenType.CROSS,
  update: TokenType.UPDATE,
  set: TokenType.SET,
};
