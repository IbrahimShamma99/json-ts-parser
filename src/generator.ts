import { JsVisitor } from "./interpreter/js.visitor";
import { Scanner } from "./scanner";
import { Parser } from "./parser";
import { Token, TokenType } from "./tokens";
import { Statement } from "./classes/statements/statement";

export class Generator {
  constructor(private data: Record<any, any>) {}

  run(sql: string) {
    const scanner = new Scanner(sql);
    scanner.scan();

    const parser = new Parser(scanner.tokens);

    const tree = parser.parse();

    const jsVisitor = new JsVisitor().execute(tree);

    const jsCode = `function run() { return ${JSON.stringify(
      this.data
    )}${jsVisitor} } run()`;

    return eval(jsCode);
  }

  scan(sql: string): Token<TokenType>[] {
    const scanner = new Scanner(sql);
    scanner.scan();
    return scanner.tokens;
  }

  parse(tokens: Token<TokenType>[]): Statement[] {
    const parser = new Parser(tokens);
    return parser.parse();
  }

  visit(statements: Statement[]): string {
    const jsVisitor = new JsVisitor().execute(statements);
    return jsVisitor;
  }

  generateCode(source: string): string {
    const jsCode = `function run() { return ${JSON.stringify(
      this.data
    )}${source} } run()`;

    return jsCode;
  }

  execute(source: string): any {
    const jsCode = this.generateCode(source);
    return eval(jsCode);
  }
}
