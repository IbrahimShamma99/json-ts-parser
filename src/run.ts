import { Scanner } from "./scanner";
import { Parser } from "./parser";
import { JsVisitor } from "./interpreter/js.visitor";
import { generate } from "./sql-js";

const run = async () => {
  const source = `
    SELECT
      id,
      user_name as username

        from users as u

        ;
      `;

  const results = generate(
    [
      { user_name: "user1", id: "id1" },
      { user_name: "user2", id: "id2" },
    ],
    source
  );

  console.log(results);
};

run();
