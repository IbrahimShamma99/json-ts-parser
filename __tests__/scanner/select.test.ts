import { Scanner } from "../../src/scanner";

describe("Scanner Tests", () => {
  it("should scan a simple select statement", () => {
    const source = `
            SELECT
                id,
                user_name as username
                from users
                `;

    const scanner = new Scanner(source);
    scanner.scan();

    expect(scanner.tokens).toMatchSnapshot("tokens");
  });
});
