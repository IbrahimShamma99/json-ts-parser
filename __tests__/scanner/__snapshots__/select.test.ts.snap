// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scanner Tests should scan a simple select statement 1`] = `
[
  Token {
    "end": 20,
    "lexeme": "select",
    "line": 2,
    "start": 14,
    "type": "SELECT",
  },
  Token {
    "end": 39,
    "lexeme": "id",
    "line": 2,
    "start": 37,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 39,
    "lexeme": "id",
    "line": 2,
    "start": 37,
    "type": "COMMA",
  },
  Token {
    "end": 66,
    "lexeme": "user_name",
    "line": 2,
    "start": 57,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 69,
    "lexeme": "as",
    "line": 2,
    "start": 67,
    "type": "AS",
  },
  Token {
    "end": 78,
    "lexeme": "username",
    "line": 2,
    "start": 70,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 99,
    "lexeme": "from",
    "line": 2,
    "start": 95,
    "type": "FROM",
  },
  Token {
    "end": 105,
    "lexeme": "users",
    "line": 2,
    "start": 100,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 121,
    "lexeme": "",
    "line": 2,
    "start": 121,
    "type": "EOF",
  },
]
`;

exports[`Scanner Tests should scan a simple select statement: tokens 1`] = `
[
  Token {
    "end": 20,
    "lexeme": "select",
    "line": 2,
    "start": 14,
    "type": "SELECT",
  },
  Token {
    "end": 39,
    "lexeme": "id",
    "line": 2,
    "start": 37,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 39,
    "lexeme": "id",
    "line": 2,
    "start": 37,
    "type": "COMMA",
  },
  Token {
    "end": 66,
    "lexeme": "user_name",
    "line": 2,
    "start": 57,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 69,
    "lexeme": "as",
    "line": 2,
    "start": 67,
    "type": "AS",
  },
  Token {
    "end": 78,
    "lexeme": "username",
    "line": 2,
    "start": 70,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 99,
    "lexeme": "from",
    "line": 2,
    "start": 95,
    "type": "FROM",
  },
  Token {
    "end": 105,
    "lexeme": "users",
    "line": 2,
    "start": 100,
    "type": "IDENTIFIER",
  },
  Token {
    "end": 121,
    "lexeme": "",
    "line": 2,
    "start": 121,
    "type": "EOF",
  },
]
`;
